### PR TITLE
Remove scanning of segment on startup

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

To reduce the startup time, this PR removes scanning of the segments on startup. Instead metadata about last entry is persisted in the descriptor. On startup, segment writer can jump to the last entry without scanning the whole segment. 

`SegmentDescriptor` now contains two more fields:
- index of the last entry
- position of the last entry
This is used by `SegmentWriter` to jump to the last entry, if present. If these meta data is not present, it switches to scanning the whole segment.

`lastIndex` and `lastPosition` in the descriptor is only updated once when the writer moves from the segment to the next. This is done so to keep it simple. It is difficult to keep update the descriptor and the last entry in sync, as it involves when the blocks are flushed. As a result, the last segment will not have these metadata in the descriptor. So after a restart, the writer has to scan the last segment. This is ok because it is just one segment, and it will be useful to built the index in that segment.

There are some potential inconsistency between updating the descriptor and the actual last entry.
1. The descriptor is updated and flushed to disk. The last entry is not flushed to disk. On startup, when the writer jump to the lastPosition, it may not find the entry and could detect it as corruption. When this happens, we scan the whole segment to reset the writer.
2. The descriptor is not updated or flushed to disk. The last entry is flushed to disk. This is ok. If the last index and position is 0, the writer scans the segment to reset. If the last index and position is not 0, but if it points to an incorrect entry, it would detect it as corruption and switch to scanning. 
3. Segment 1 updates last index and position, write to descriptor, and move to Segment 2.  Then the journal is truncated to an index in Segment 1. Segment 2 is deleted. The entry representing last index and position in the descriptor is also deleted. This could lead to inconsistency after restart, because the writer might jump to the last entry and may not detect it as deleted or corrupted. To prevent this, before truncation we update the descriptor and flush to disk immediately.

Another caveat is that, if the descriptor was written using the old version, we should never overwrite it. This is because the descriptor length changes and this could overflow into the first entry and overwrite it. 

Note:- In this PR, after restart the index of existing segments are never re-built. This will be done in a follow up PR. 

## Related issues

related #12667 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [x] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by deepthidevaki** [comment]
@npepinpe I have assigned you as reviewer. We haven't discussed this approach before. Let me know if you would prefer a walk though before.

Benchmark is running under namespace `dd-12667-segment-descriptor-benchmark`. I will update the startup time, once it fills up the disk.

**Review by npepinpe — APPROVED** [review]
:rocket:

- :wrench: We could also update the descriptor on graceful shutdown to avoid having to scan the last segment on restart. Or do we want to keep the semantic meaning of `lastIndex` and `lastPosition` to really mean the very last, and not possibly some intermediate index/position? I think it wouldn't matter, but let me know.
- :upside_down_face: Slightly out of scope, but to avoid issues with changing the descriptor in the future, we could either: 1) write the descriptor at the end of the file, allowing it to grow, or 2) reserve more space for the descriptor, e.g. 512 bytes or something sufficiently larger

File: journal/src/main/java/io/camunda/zeebe/journal/file/Segment.java
Comment (deepthidevaki) [reply journal/src/main/java/io/camunda/zeebe/journal/file/Segment.java:292]: This doesn't reduce the guarantees, right? It might flush more often than we expected. So in that sense it is not reducing the guarantee. 
Besides, this is only done during truncation, which will not happen often.

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentWriter.java
Comment (deepthidevaki) [reply journal/src/main/java/io/camunda/zeebe/journal/file/SegmentWriter.java:310]: I feel that persisting index is not worth it, because we will delete the segments as fast as we can. So persisting the index just to delete it a few minutes later sounds not useful.

**Comment by deepthidevaki** [comment]
> * wrench We could also update the descriptor on graceful shutdown to avoid having to scan the last segment on restart. Or do we want to keep the semantic meaning of `lastIndex` and `lastPosition` to really mean the very last, and not possibly some intermediate index/position? I think it wouldn't matter, but let me know.

We can update the descriptor on graceful shutdown. But I think it is unnecessary. I think it is ok scan the last segment and built its index on startup. It is likely that atleast one reader has to seek inside this segment. 

> * upside_down_face Slightly out of scope, but to avoid issues with changing the descriptor in the future, we could either: 1) write the descriptor at the end of the file, allowing it to grow, or 2) reserve more space for the descriptor, e.g. 512 bytes or something sufficiently larger

That makes sense. Moving descriptor to the end would be difficult to implement in a backward compatible way. Currently, the descriptor also contains the segment file size, which I think we use it to map the buffer. Not sure if it creates any issue if we move the descriptor to the end.

Reserving a fixed space for descriptor would be relatively easy. I could give it a try.

**Comment by deepthidevaki** [comment]
> Reserving a fixed space for descriptor would be relatively easy. I could give it a try.

This is a bit more involved that I thought. I will try it out after merging this PR.

@npepinpe What are your thoughts on backporting this?

**Comment by npepinpe** [comment]
I think it's worth back porting. We can consider not back porting it to older versions if it requires a significant manual effort, but it's pretty valuable.

**Comment by deepthidevaki** [comment]
I will backport this once the follow up PR is also merged, to ensure that no partial changes are included in patch release.

**Comment by deepthidevaki** [comment]
bors merge

**Comment by deepthidevaki** [comment]
RollingUpdate test is consistency failing, not flaky. Will fix it before merging.

**Comment by deepthidevaki** [comment]
```
+----------------------------------------------------------------------------------------------------------------------
| frame |  sbe_headers of metadata | checksum | length | sbe_headers of Data | index | asqn | length | <raft-entry> |
+----------------------------------------------------------------------------------------------------------------------
```
Above is the record format. Checksum includes sbe_headers of Data, index, asqn, length and the raft-entry. During replication leader sends the `PersistedRaftRecord` which includes checksum, index, asqn, and raft-entry. Note:- It doesn't send include the sbe_headers of Data. When follower receives it and write to the the journal, it writes using the sbe serializer and generates sbe_headers of Data. If the follower and leader are on different versions (sbe version of journal schema), it would result in different checksum. :face_palm: This is why RollingUpdateTest fails.

This is a major blocker in making changes to journal sbe schema. This PR is blocked until we find a solution to this.

**Comment by npepinpe** [comment]
:exploding_head:

**Comment by deepthidevaki** [comment]
@npepinpe I rebased this PR to main to include the fix for #12915. There was small merge conflicts in SegmentedJournalWriter and in some tests. But nothing much that require a second review. After rebasing, I added following changes ([see diff](https://github.com/camunda/zeebe/compare/0a272e7...0980ac3f5)) for which review would be nice.
- https://github.com/camunda/zeebe/pull/12839/commits/538d1390c4fec987caea54231827e5f93257e5c0 
- https://github.com/camunda/zeebe/pull/12839/commits/90bcd37eeb884a7cd09486240f6866b083b920fb
- https://github.com/camunda/zeebe/pull/12839/commits/39567b961e417abd2e88d970dacb7a304096e3ac
- https://github.com/camunda/zeebe/pull/12839/commits/0980ac3f52566c00ce9f6c48746ba97a8268e20c

**Review by npepinpe — APPROVED** [review]
:question: All the old append paths can be removed only by 8.4, right?

File: journal/src/main/java/io/camunda/zeebe/journal/file/Segment.java
Comment (deepthidevaki) [reply journal/src/main/java/io/camunda/zeebe/journal/file/Segment.java:292]: True. It could be confusing. But I think we also force flush when we take snapshot, overriding the configured flush. Sometimes it is necessary to prevent inconsistencies.

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentDescriptor.java
Comment (deepthidevaki) [reply journal/src/main/java/io/camunda/zeebe/journal/file/SegmentDescriptor.java:58]: It is only used when loading the segment to verify if the file is large enough to store the descriptor, I think. I can take a look if it is ok to remove it.

**Comment by deepthidevaki** [comment]
> All the old append paths can be removed only by 8.4, right?

Yes. But we can already remove them after 8.3.0 release. It will be live in 8.4 only.

File: journal/src/main/java/io/camunda/zeebe/journal/file/SegmentDescriptor.java
Comment (deepthidevaki) [reply journal/src/main/java/io/camunda/zeebe/journal/file/SegmentDescriptor.java:58]: Created follow up for this and using fixed length descriptor https://github.com/camunda/zeebe/issues/13240

**Pull request merged**
